src/System.Linq/src/System/Linq/Select.SpeedOpt.cs

```diff
@@ -157,6 +157,127 @@ public TResult TryGetLast(out bool found)
             }
         }
 
+        private sealed partial class SelectRangeIterator<TResult> : Iterator<TResult>, IPartition<TResult>
+        {
+            private readonly int _start;
+            private readonly int _end;
+            private readonly Func<int, TResult> _selector;
+
+            public SelectRangeIterator(int start, int end, Func<int, TResult> selector)
+            {
+                Debug.Assert(start < end);
+                Debug.Assert((end - start) <= int.MaxValue);
+                Debug.Assert(selector != null);
+
+                _start = start;
+                _end = end;
+                _selector = selector;
+            }
+
+            public override Iterator<TResult> Clone() =>
+                new SelectRangeIterator<TResult>(_start, _end, _selector);
+
+            public override bool MoveNext()
+            {
+                if (_state < 1 || _state == (_end - _start + 1))
```

Comment:
> it would be great if the base class (Iterator) was exposing sth like

IsDisposed would be `_state == -1`, and that's not actually the condition being checked here.

> why the check is not _state < 0 instead of _state < 1?

_state starts out at 0 in the ctor.  If you call MoveNext and it's in that state, you've messed up and MoveNext bails immediately.  You need to call GetEnumerator on the instance, which sets _state to 1, at which point it's valid to use MoveNext.  Essentially this condition is checking for someone erroneously doing this:
```C#
IEnumerator<int> e = (IEnumerator<int>)Enumerable.Select(...);
e.MoveNext();
```

Reply:
@stephentoub thanks for the explanation!